lib/PuppeteerSharp/WaitTask.cs

```diff
@@ -183,84 +87,127 @@ public void Dispose()
+                            }",
+                            new object[]
+                            {
+                            await _isolatedWorld.GetPuppeteerUtilAsync().ConfigureAwait(false),
+                            _root,
+                            _fn,
+                            }.Concat(_args).ToArray()).ConfigureAwait(false);
                 }
 
-                return;
-            }
+                await _poller.EvaluateFunctionAsync("poller => poller.start()").ConfigureAwait(false);
 
-            if (exception == null &&
-                await _isolatedWorld.EvaluateFunctionAsync<bool>("s => !s", success)
-                    .ContinueWith(
-                        task => task.IsFaulted || task.Result,
-                        TaskScheduler.Default)
-                    .ConfigureAwait(false))
+                var success = await _poller.EvaluateFunctionHandleAsync("poller => poller.result()").ConfigureAwait(false);
+                _result.TrySetResult(success);
+                await TerminateAsync().ConfigureAwait(false);
+            }
+            catch (Exception ex)
             {
-                if (success != null)
+                var exception = GetBadException(ex);
+                if (exception != null)
                 {
-                    await success.DisposeAsync().ConfigureAwait(false);
+                    await TerminateAsync(exception).ConfigureAwait(false);
                 }
-
-                return;
             }
+        }
 
-            if (exception?.Message.Contains("Execution context was destroyed") == true)
+        internal async Task TerminateAsync(Exception exception = null)
+        {
+            // The timeout timer might call this method on cleanup
+            if (_terminated)
             {
-                _ = Rerun();
                 return;
             }
 
-            if (exception?.Message.Contains("Cannot find context with specified id") == true)
-            {
-                return;
-            }
+            _terminated = true;
+            _isolatedWorld.TaskManager.Delete(this);
+            Cleanup(); // This matches the clearTimeout upstream
 
             if (exception != null)
             {
-                _taskCompletion.TrySetException(exception);
+                _result.TrySetException(exception);
             }
-            else
+
+            if (_poller != null)
             {
-                _taskCompletion.TrySetResult(success);
-            }
+                try
+                {
+                    await _poller.EvaluateFunctionAsync(@"async poller => {
+                        await poller.stop();
+                    }").ConfigureAwait(false);
 
-            Cleanup();
+                    await _poller.DisposeAsync().ConfigureAwait(false);
+                    _poller = null;
+                }
+                catch (Exception)
+                {
+                    // swallow error.
+                }
+            }
         }
```

Comment:
I'm not into the execution flow of when `Rerun` is executed, but I just wanted to point out a _potential_ race-condition.
If one thread running `Rerun` assigns a new value to `_poller` while another thread is in here, we might invoke `EvaluateFunctionAsync` and `DisposeAsync` on different instances.

So my suggestion is to save `_poller` to a local variable and act on that instead of the mutable field.

E.g. 
```cs
if (_poller is { } poller)
{
    try
    {
        await poller.EvaluateFunctionAsync(@"async poller => {
            await poller.stop();
        }").ConfigureAwait(false);

        await poller.DisposeAsync().ConfigureAwait(false);
        _poller = null;
    }
    catch (Exception)
    {
        // swallow error.
    }
}
```

Reply:
To put our suggestions into code
```cs
if (_poller is { } poller)
{
    _poller = null;

    await using (poller.ConfigureAwait(false))
    {
        try
        {
            await poller.EvaluateFunctionAsync(@"async poller => {
                await poller.stop();
            }").ConfigureAwait(false);
        }
        catch (Exception)
        {
            // swallow error.
        }
    }
}
```

Reply:
That's interesting. It's quite an edge scenario the best thread-safe we're, the better. 
Thank you for this idea!